data_structures/treap.cpp

```diff
@@ -0,0 +1,228 @@
+     */
+    void _insert(int &x, int k) {
+        if (x) {
+            if (key[x] == k) {
+                cnt[x]++;
+            }  // If the node already exists, the number of copies is ++
+            else {
+                int t = (key[x] < k);  // Insert according to BST properties
+                _insert(childs[x][t], k);
+                // After insertion, the heap properties are retained by rotation
+                if (priority[childs[x][t]] < priority[x]) {
+                    rotate(x, t);
+                }
+            }
+        } else {  // Create a new node
+            x = treapCnt++;
+            key[x] = k;
+            cnt[x] = 1;
+            priority[x] = rand();  // Random priority
+            childs[x][0] = childs[x][1] = 0;
+        }
+        update(x);
+    }
+    /**
+     * @brief Erase internal method
+     * @param x Erase from the subtree of node x (Usually x=root)
+     * @param k key
+     */
+    void _erase(int &x, int k) {
+        if (key[x] == k) {
+            if (cnt[x] > 1) {
+                cnt[x]--;
+            }  // If the node has more than one copy, the number of copies --
+            else {
+                if (childs[x][0] == 0 && childs[x][1] == 0) {
+                    x = 0;
+                    return;
+                }  // If there are no children, delete and return
+                // Otherwise, we need to rotate the sons and delete them
+                // recursively
+                int t = (priority[childs[x][0]] > priority[childs[x][1]]);
+                rotate(x, t);
+                _erase(x, k);
+            }
+        } else {  // Find the target value based on BST properties
+            _erase(childs[x][key[x] < k], k);
+        }
+        update(x);
+    }
+    /**
+     * @brief Find the KTH largest value (Also internal method)
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried rank
+     * @return The element ranked x
+     */
+    int _getKth(int &x, int k) {
+        if (k <= size[childs[x][0]]) {
+            return _getKth(childs[x][0], k);
+        }
+        k -= size[childs[x][0]] + cnt[x];
+        if (k <= 0) {
+            return key[x];
+        }
+        return _getKth(childs[x][1], k);
+    }
+    /**
+     * @brief Query the rank of specified element (Also internal method)
+     * @param x Query the subtree of node x (Usually x=root)
+     * @param k The queried element
+     * @return The rank of element k
+     */
+    int _getRank(int x, int k) {
+        if (!x)
+            return 0;
+        if (k == key[x])
+            return size[childs[x][0]] + 1;
+        else if (k < key[x])
+            return _getRank(childs[x][0], k);
+        else
+            return size[childs[x][0]] + cnt[x] + _getRank(childs[x][1], k);
+    }
+    /**
+     * @brief Get the predecessor node of element k
+     * @param k The queried element
+     * @return The predecessor
+     */
+    int getPre(int k) {
```

Comment:
I think the function names should be fully typed out so there is no confusion.
the standards requires snake case for function names
 ```suggestion
    int get_predecessor(int k) {
```